File: P4924.c
#include <stdio.h>

#define IVS_N 505

int matrix[IVS_N][IVS_N];
int newMatrix[IVS_N][IVS_N];

void ivsRotate(int ii, int ij, int ir, int id) {
    int len = 2 * ir + 1;
    int begI = ii - ir, begJ = ij - ir, endI = ii + ir, endJ = ij + ir;

    if (id == 0)
        for (int i = 0; i < len; ++i)
            for (int j = 0; j < len; ++j)
                newMatrix[begI + j][endJ - i] = matrix[begI + i][begJ + j];
    else
        for (int i = 0; i < len; ++i)
            for (int j = 0; j < len; ++j)
                newMatrix[endI - j][begJ + i] = matrix[begI + i][begJ + j];

    for (int i = begI; i <= endI; ++i)
        for (int j = begJ; j <= endJ; ++j)
            matrix[i][j] = newMatrix[i][j];
}

int main() {
    int n, m;
    scanf("%d%d", &n, &m);

    int v = 1;
    for (int i = 1; i <= n; ++i)
        for (int j = 1; j <= n; ++j) {
            matrix[i][j] = v;
            ++v;
        }

    int a, b, c, d;
    while ((scanf("%d%d%d%d", &a, &b, &c, &d)) != EOF)
        ivsRotate(a, b, c, d);

    for (int i = 1; i <= n; ++i) {
        for (int j = 1; j <= n; ++j)
            printf("%d ", matrix[i][j]);
        putchar('\n');
    }

    return 0;
}

File: P4924.java
import java.util.Scanner;

public class Main {
    private static void rotate(int i, int j, int r, int d) {
        int len = 2 * r + 1;
        int begI = i - r, begJ = j - r, endI = i + r, endJ = j + r;

        if (d == 0)
            for (int x = 0; x < len; ++x)
                for (int y = 0; y < len; ++y)
                    newMatrix[begI + y][endJ - x] = matrix[begI + x][begJ + y];
        else
            for (int x = 0; x < len; ++x)
                for (int y = 0; y < len; ++y)
                    newMatrix[endI - y][begJ + x] = matrix[begI + x][begJ + y];

        for (int t = begI; t <= endI; ++t)
            System.arraycopy(newMatrix[t], begJ, matrix[t], begJ, len);
    }

    public static void main(String[] args) {
        Scanner ivsScan = new Scanner(System.in);
        int n = ivsScan.nextInt();
        ivsScan.nextInt();

        matrix = new int[n + 2][n + 2];
        int x = 1;
        for (int i = 1; i <= n; ++i)
            for (int j = 1; j <= n; ++j) {
                matrix[i][j] = x;
                ++x;
            }

        newMatrix = new int[n + 2][n + 2];

        while (ivsScan.hasNextInt()) {
            int a = ivsScan.nextInt();
            int b = ivsScan.nextInt();
            int c = ivsScan.nextInt();
            int d = ivsScan.nextInt();
            rotate(a, b, c, d);
        }

        for (int i = 1; i <= n; ++i) {
            for (int j = 1; j <= n; ++j) {
                System.out.print(matrix[i][j]);
                System.out.print(' ');
            }
            System.out.println();
        }
    }

    static private int[][] matrix;
    static private int[][] newMatrix;
}
